File: ts/640/589.ts
/**
 * Definition for node.
 */
class NTreeNode {
  val: number
  children: NTreeNode[]
  constructor(val?: number) {
    this.val = (val===undefined ? 0 : val)
    this.children = []
  }
}

function preorder(root: NTreeNode | null): number[] {
  const ans: number[] = []
  const preorder = (node: NTreeNode | null): void => {
    if (!node) return
    ans.push(node.val)
    for (const child of node.children) {
      preorder(child)
    }
  }
  preorder(root)
  return ans
};

File: c/640/589.c
#include <stdio.h>
#include <stdlib.h>

/**
 * Definition for a Node.
 */
struct Node {
  int val;
  int numChildren;
  struct Node** children;
};

void dfs(struct Node* node, int* res, int* pos) {
  if (NULL == node) return;
  res[(*pos)++] = node->val;
  for (int i = 0; i < node->numChildren; ++i) {
    dfs(node->children[i], res, pos);
  }
}

int* preorder(struct Node* root, int* returnSize) {
  int* ans = (int*)malloc(sizeof(int) * 10000); // 10000 题目所给树的结点总数
  int i = 0;
  dfs(root, ans, &i);
  *returnSize = i;
  return ans;
}